### ISSUE TITLE
Crash on unit destructor, related to AuraEffects list

### ISSUE THREAD

**Issue by jerony-mo** [TrinityCore/TrinityCore#26495]
**Description:**

Server crashes min. once a day on this one. Seems to be related to AuraEffects on unit destructor. 
I found really old similar issue: https://github.com/TrinityCore/TrinityCore/issues/18028
Server has a couple of custom fixes, but nothing extravagant. 

Crashlog: https://pastebin.com/65wstE3g

**Expected behaviour:**

Should not crash.

**Steps to reproduce the problem:**

No idea

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

6e36a2007cc64a4fe11cc7781c9ca6f619118a20

**Operating system:** 

Debian 4.19

**Comment by illfated** [comment]
Looks like there is quite a difference between the line numbers in the TC source and the lines referenced in your crash log. You may need to use gdb on debug build servers, if not also a clean TrinityCore source, to produce a useful crash log to make it possible for the TC members to find the exact cause of this crash.

(This is only my initial thoughts while trying to match your log entries to the TC source. Maybe a real coder will be able to find something more substantial.)

**Comment by jerony-mo** [comment]
Yeah, I've posted it here just in case anyone has any idea what may be the cause. Unfortunately, not only the line numbers doesn't match current official core (that's normal as we have some changes there), but line Unit.cpp:422 actually points out to main closing bracket `}` on Unit destructor function which effectively is useless.

We have a full core dump available for gdb, unfortunately it's only in RelWithDebInfo build. If no one has any idea, I'd have to sacrifice performance to build and run in Debug mode.

**Comment by Shauren** [comment]
Line number pointing at `}` is very much a normal thing here - if you notice its crashing inside a destructor of std::list of aura effects, it is called automatically without writing code to do it

**Comment by jackpoz** [comment]
If it happens once a day you could try enabling Address Sanitizer and getting a better log.

Do you have a public fork with the custom changes ?

**Comment by Idoitbetter** [comment]
I've got similar crash: https://gist.github.com/Idoitbetter/f5da84cd5bda9a876cbdaed30debab73
https://github.com/TrinityCore/TrinityCore/commit/f418631de0c193772e24f379db5aafcc01f3fb6a

Also seems similar to this old issue: https://github.com/TrinityCore/TrinityCore/issues/234

And I found the same crash here: https://github.com/jackpoz/TrinityCore/issues/75#issuecomment-506925042

I don't know how to reproduce it, Is there any way to add some log to get some more information, possible auras or something?

**Comment by Idoitbetter** [comment]
this crash continues to occur regularly (with many players online):
https://gist.github.com/Idoitbetter/a1d1c7a70c74137ea1fff870852b94db
f9033a5dbd559fde3030a21d83d664983d95f39f

Can happens in any instance, I added a log inside DestroyInstance to see map instance id, and It is not one in particular, it can be anywhere

Does anyone know what's going on or a way to locate the problem?, I don't know what happens with m_modAuras, maybe memory leak or bad data? or due to multithreading when creature removes from world when instance is destroyed?

I have added bountysource in case anyone is encouraged

**Comment by Idoitbetter** [comment]
I added a log in Unit::~Unit()
<details>
  <summary>Log</summary>

```c++
    for (uint32 i = 0; i < TOTAL_AURAS; ++i)
    {
        if (!m_modAuras[i].empty())
        {
            std::stringstream sstr;
            sstr << "Test m_modAuras: Unit Entry: " << GetEntry() << " MapId: " << GetMapId() << " m_modAurasSize: " << m_modAuras[i].size() << " itr: " << i;
            TC_LOG_FATAL("entities.unit.destructor", "%s", sstr.str().c_str());
            if (m_modAuras[i].size() > 0)
            {
                for (AuraEffectList::iterator iter = m_modAuras[i].begin(); iter != m_modAuras[i].end(); ++iter)
                {
                    uint32 AuraID = ((*iter) && (*iter)->GetSpellInfo()) ? (*iter)->GetSpellInfo()->Id : 0;
                    sstr << " AuraID: " << AuraID << " \n";
                }
            }
            TC_LOG_FATAL("entities.unit.destructor", "%s", sstr.str().c_str());
        }
    }
```

</details>

And I only got data just when crash happens
First time:
```
Test m_modAuras: Unit Entry: 32593 MapId: 601 m_modAurasSize: 0 itr: 261
```
Second:
```
Test m_modAuras: Unit Entry: 37496 MapId: 658 m_modAurasSize: 0 itr: 197
Test m_modAuras: Unit Entry: 37496 MapId: 658 m_modAurasSize: 0 itr: 197
```
It doesn't make sense those npc with that aura types, I also can't understand how it could be that m_modAuras is not empty, but its size is 0.

**Comment by Shauren** [comment]
> I also can't understand how it could be that m_modAuras is not empty, but its size is 0.

empty() is not implemented as size() == 0 (but this doesnt matter, the list got corrupted externally)

**Comment by Idoitbetter** [comment]
Just in case it helps, this has been appearing in the log when the crash occurs:

```
Test m_modAuras: Unit Entry: 2110 MapId: 658 m_modAurasSize: 0 itr: 101

Test m_modAuras: Unit Entry: 29630 MapId: 604 m_modAurasSize: 0 itr: 189

Test m_modAuras: Unit Entry: 17962 MapId: 547 m_modAurasSize: 0 itr: 157
```

The last one is intriguing, as aura type 157 is associated with only one spell, which is the hunter's pet (passive). This doesn't make sense for that NPC. I don't understand how this could have happened or if the m_modAuras array is also corrupted due to the existing corruption.

I don't know how to figure out how this can happen. Is it possible that when running _RegisterAuraEffect(AuraEffect* aurEff, bool apply) due to multithreading, it could be causing the problem? Would adding a std::lock_guard there help prevent it, or would it have no effect? I don't know what else to do.

**Comment by jackpoz** [comment]
You could enable ASan and get a report from that

**Comment by Idoitbetter** [comment]
Hello, thanks for the advice, I have ASan in a test realm where a few users enter to do many tests, but unfortunately I cannot put it in a realm where many players enter, performance problems immediately begin and it is not possible.

**Comment by jackpoz** [comment]
did you build in RelWithDebInfo or ?

**Comment by Idoitbetter** [comment]
Yes, I always compile in RelWithDebInfo
